### ISSUE TITLE
Modern version with CSS variables

### ISSUE THREAD

**Issue by jonaskuske** [kognise/water.css#70]
So, turning this into a separate issue instead of just a [comment](https://github.com/kognise/water.css/issues/25#issuecomment-489369081) in #25:

I'd like to see a "modern" version of `water.css`, using actual CSS variables (aka Custom Properties) for theming. They're supported in all modern browsers for a little more than two years now, [here's the full browser support sheet](https://caniuse.com/#feat=css-variables).

This would allow us to ship builds including *both* the light and the dark theme (without having to load two separate files) for devs who want to add both themes to their site, to make theming depend on users' system-wide theme preference. (see #25)

We could still transpile the variables to static CSS where appropriate so we wouldn't lose any browser support, just add additional smaller builds for those who don't have to support IE any longer.

**This would give us the following builds:**
(you can also explore the options here: https://codepen.io/21stCenturyJonas/full/JqjPgL)

### `light.standalone.css`
Includes light variables and core, compiled to "normal" static CSS – same as current `light.css`:
```css
@import 'parts/_variables-light.scss';
@import 'parts/_core.scss';
```

### `dark.standalone.css`
Includes dark variables and core, compiled to "normal" static CSS – same as current `dark.css`:
```css
@import 'parts/_variables-dark.scss';
@import 'parts/_core.scss';
```

### `dark.css`
Includes both dark and light variables, left as CSS custom properties so they can be changed dynamically. Dark is the default, light variables are only applied if `(prefers-color-scheme: light)` matches.
```css
@import 'parts/_variables-dark.scss';
@media (prefers-color-scheme: light) {
  @import 'parts/_variables-light.scss';
}
@import 'parts/_core.scss';
```

### `light.css`
Includes both light and dark variables, left as CSS custom properties so they can be changed dynamically. Light is the default, dark variables are only applied if `(prefers-color-scheme: dark)` matches.
```css
@import 'parts/_variables-light.scss';
@media (prefers-color-scheme: dark) {
  @import 'parts/_variables-dark.scss';
}
@import 'parts/_core.scss';
```

### `dark-legacy.css`
URL-imports (→ at runtime in the browser) the compiled, static dark theme and, in case `(prefers-color-scheme: light)` matches, the compiled, static light theme, too. Requires multiple network requests and a larger download size overall but enables user-preference dependent theming in legacy browsers.
```css
@import url('dark.standalone.css');
@import url('light.standalone.css') (prefers-color-scheme: light);
```

### `light-legacy.css`
URL-imports (→ at runtime in the browser) the compiled, static light theme and, in case `(prefers-color-scheme: dark)` matches, the compiled, static dark theme, too. Requires multiple network requests and a larger download size overall but enables user-preference dependent theming in legacy browsers.
```css
@import url('light.standalone.css');
@import url('dark.standalone.css') (prefers-color-scheme: dark);
```

<hr>

Happy to work on this once mission is a go 😀

**Comment by jonaskuske** [comment]
Tasks to implement this would be (as far as I can see):

- [ ] Rewrite variable usage from Sass syntax ( `$my-variable: value` / `property: $my-variable` ) to CSS syntax ( `--my-variable: value` / `property: var(--my-variable)` )
- [ ] Create partials for `_variables-light.scss` and `_variables-dark.scss` in `parts/`
- [ ] Add the build files described above (with correct import links)
- [ ] Update build step to compile vars for `standalone` & `legacy` files (→ PostCSS, subdependency already)
- [ ] Add separate `browserslist` config so we don't add prefixes in the modern builds even though they're unnecessary in all browsers supporting CSS variables (→ further reducing file size)
- [ ] Add or update installation instructions on the demo site
- [ ] Explain CSS variable usage in Contributing guide (?)

**Comment by kylejrp** [comment]
Oh wow, I should have read this before making my comment in https://github.com/kognise/water.css/issues/25#issuecomment-489389211. Sounds like you've already have a good grasp on the work that needs to be done, but give me some time to review what you've got first. (Great research and work! 😄)

**Comment by kylejrp** [comment]
Okay, I love it. I think you should go ahead and run with this.

My only concerns are:
1. How does this affect installation?
2. How does this affect future code contributors?
3. How does this affect future theme contributions?

I think 1. is solved by the new UI you've mocked up, especially with the advanced options feature.

I think 2. is my bigger concern. I'm worried about the root of `/src/` getting a little messy if we have a few `.scss` configurations. It might make sense to create a new folder `/src/builds` or `/src/configurations` that stores all of the build configurations you've listed above. I think putting it in a folder will hide our build configurations from our users. Maybe we just need clear directions in the Contributing section.

For 3, I think it would be important to list which files that theme contributors will have to modify/import to create a new theme in the Theming section. It's no longer defining variables in one file in one location, it's modifying variables and several build files in a couple of locations.

Finally, I think switching from Sass variables to CSS variables is a breaking change and will require an appropriate semantic version change. (cc @kognise)

**Comment by jonaskuske** [comment]
Replying to your comment here:

> The trickiest part here would be modifying our existing Sass/Gulp to handle this. Ideally, we just have Sass variables or CSS variables, not a mix of both. Here's a few options:
> 1. Switch all code to use CSS variables, use https://www.npmjs.com/package/postcss-css-variables to "compile" these when necessary for browser support
> 2. Switch all code to use CSS variables, offer the CSS variables polyfill for browser support
> 3. Keep Sass variables, compile these to CSS variables somehow (?)
> 
> I think 1 makes the most sense here, but there's some work to get this set up. I can try making a pull request that uses the CSS variables while still supporting legacy browsers where necessary.

The CSS variable polyfill is super cool, but it also is a big chunk of JavaScript that has to download and parse every single stylesheet on a page, compile it and insert it after pageload. It's very much impressive but also very much overkill for our usecase. 😀 Using Sass variables as custom property works ( `--css-var: $sass-var;` and done), but you're right that that's kinda messy, so I'd favor a PostCSS-based solution, too.

We should also consider having three sets of variables: light, dark and common (or "shared", "base", ...)
This way builds including both the dark and light theme won't end up with common variables like timings twice. (we could also compile base variables for *every* build, as they never need to be dynamic at runtime unless I'm missing something)

<hr>

Regarding your concerns: I think installation should be fine, but yep, Contributing and Theming sections will definitely need some love and we'll need a good code base structure backing this change.

As for the breaking change: yeah, since we're actively advertising 3rd party theming as part of our "public API" that'd be a breaking change I guess. Even more so, changing the default `dark.css`/`light.css` styles to reflect `(prefers-color-scheme)` (while dropping legacy browser support) is **definitely** a breaking change. We could avoid that by making my proposed `standalone` builds the defaults, though. But I'm personally very much in favor of changing our defaults as I feel that taking this hands-off approach, following user preference as much as possible and simply making sure the page looks good whatever the circumstances fits in very well with the "water.css philosophy".

**Comment by kylejrp** [comment]
> It's very much impressive but also very much overkill for our usecase. ... I'd favor a PostCSS-based solution, too.

Agreed!

> We should also consider having three sets of variables: light, dark and common (or "shared", "base", ...)

Sure, I can see this being useful, as long as they're overwrite-able when necessary. (Theme contributors?)

> we could also compile base variables for every build

Do you mean that the common variables would remain SASS variables and compiled into the CSS? I think that makes sense. I think we only need to use CSS variables when they're being switched from light/dark.

> I'm personally very much in favor of changing our defaults as I feel that taking this hands-off approach, following user preference as much as possible and simply making sure the page looks good whatever the circumstances fits in very well with the "water.css philosophy".

I definitely agree. I'm in favor of including `(prefers-color-scheme)` in the default and just ticking up the version number as a breaking change. It's a reasonable default and it's not that big of a deal to include.

In any case, you've got an awesome understanding of what's involved with this issue and I'm looking forward to your contribution here! Awesome work, @jonaskuske.

**Comment by jonaskuske** [comment]
Oh, as for that:

> For 3, I think it would be important to list which files that theme contributors will have to modify/import to create a new theme in the Theming section. It's no longer defining variables in one file in one location, it's modifying variables and several build files in a couple of locations.

When using the modern build, you wouldn't have to modify any of our files, you could simply define your own variables in *your* CSS, and as long as your stylesheet appears after `water.css` in the HTML, the CSS cascade will take care of everything.

To build your own full-blown theme including compilation, legacy builds and potential redistribution, you'd also create your own stylesheet containing your variables, but instead of adding it to your HTML you'd have to add it to our PostCSS plugin. The plugin takes an array of files containing variable definitions, and for every variable defined there, occurrences referencing this variable will be replaced with the actual variable value in the compiled CSS. So in our example, that config (`postcss.config.js` in the root of the repo) would look something like:

```js
{
  'postcss-css-variables': {
    imports: [
      'src/parts/_variables-base.css',
      'src/parts/_variables-dark.css',
      'src/parts/_variables-light.css',
    ]
  }
}
```

To use their own theming and override our defaults, users could simply add the path to their own file containing variables at the end of the array. (or remove the files they don't want)

> Do you mean that the common variables would remain SASS variables and compiled into the CSS? I think that makes sense. I think we only need to use CSS variables when they're being switched from light/dark.

I'm not sure if I'd leave them as SASS variables – I think having one way to define and consume variables across the entire code base is more coherent and also friendlier for external theming/contribution – but that can certainly be discussed. :)
But yes, I mean we could compile base variables, even in modern builds. So while `color: var(--text-color)` will also be `color: var(--text-color)` in the outputted CSS so that the color can change according to current theme at runtime, we could inline the values of base variables and output `transition: all 300ms` instead of `transition: all var(--timing)`.
I'm not really sure about this though. I *think* dynamic variables can have some issues when used in combination with transitions & animations in some browsers and in that case we clearly need to compile them. But by doing that, users lose the ability to change their value dynamically from within the HTML (as described at the start of this comment) so if we can avoid compiling them, maybe we should.
We'll likely require some testing before making that decision either way.

**Comment by kylejrp** [comment]
> When using the modern build, you wouldn't have to modify any of our files, you could simply define your own variables in your CSS, and as long as your stylesheet appears after water.css in the HTML, the CSS cascade will take care of everything.

Of course, that makes so much sense. I think that makes theming _more_ accessible, since you wouldn't need to mess with the build tools at all.

> To build your own full-blown theme including compilation, legacy builds and potential redistribution...

This is probably a little extreme, I imagine most themes will be okay without all the legacy options.

>I think having one way to define and consume variables across the entire code base is more coherent and also friendlier for external theming/contribution

Makes sense, this is a definite plus.

>  I think dynamic variables can have some issues when used in combination with transitions & animations in some browsers...

Do you have any sources on this? Or I guess we just test it after.

**Comment by jonaskuske** [comment]
Heh, didn't know we have this nifty deploy preview configured!
Open https://deploy-preview-71--watercss.netlify.com/ in Safari or in a dev version of Firefox, switch your system from dark to light mode and vice versa and see the magic happen :D

**Comment by jonaskuske** [comment]
This can be closed once the current master is published as a release. (version 2.0.0 I guess)

**Comment by kylejrp** [comment]
I'm closing this as the changes have been merged into master and we're just waiting on the release of 2.0.0.
